### ISSUE TITLE
[package] nspr/4.29: apple silicon/x86 cross compilation broken with conan-1.33

### ISSUE THREAD

**Issue by timblechmann** [conan-io/conan#8556]
according to coworkers, the update to conan-1.33 broke the x86_64-to-arm64 cross compilation of this package.

the nspr project builds a (native) tool to generate some sources, which will now also be compiled with the cross compilation flags.

### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **nspr/4.29**
  * Operating System+version: **MacOS**
  * Compiler+version: **Clang**
  * Conan version: **conan 1.33.0**

### Conan profile (output of `conan profile show default` or `conan profile show <profile>` if custom profile is in use)
```
[settings]
Configuration (profile_host):
[settings]
arch=armv8
arch_build=x86_64
build_type=Debug
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=12.0
os=Macos
os.version=10.11
os_build=Macos
[options]
[build_requires]
[env]
CFLAGS=-mmacosx-version-min=10.11 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk 
CXXFLAGS=-mmacosx-version-min=10.11 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk 
SDKROOT=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk
Configuration (profile_build):
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=apple-clang
compiler.libcxx=libc++
compiler.version=12.0
os=Macos
os_build=Macos
[options]
[build_requires]
[env]
```

### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```
 > ./configure '--with-mozilla' '--disable-cplus' '--disable-64bit' '--enable-strip' '--enable-debug' '--prefix=/Users/jkc/.conan/data/nspr/4.29/_/_/package/dedc51991a586e8d7ffa22d51136e70e21a48c76' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libexecdir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' '--datarootdir=${prefix}/share' --build=x86_64-apple-darwin --host=aarch64-apple-darwin
checking build system type... x86_64-apple-darwin
checking host system type... aarch64-apple-darwin
checking target system type... aarch64-apple-darwin
checking for whoami... /usr/bin/whoami
checking for aarch64-apple-darwin-gcc... no
checking for gcc... gcc
configure: WARNING: using cross tools not prefixed with host triplet
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... yes
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for aarch64-apple-darwin-ranlib... no
checking for ranlib... ranlib
checking for as... /usr/bin/as
checking for ar... /usr/bin/ar
checking for ld... /usr/bin/ld
checking for strip... /usr/bin/strip
checking for windres... no
checking how to run the C preprocessor... gcc -E
checking for gcc -pipe support... no
checking for pragma diagnostic... no
checking whether C compiler supports -fprofile-generate... yes
checking for visibility(hidden) attribute... no
checking for perl5... no
checking for perl... /usr/bin/perl
./configure: line 6444: test: darwin: integer expression expected
checking crt_externs.h usability... yes
checking crt_externs.h presence... yes
checking for crt_externs.h... yes
checking whether gcc -arch arm64 needs -traditional... rm: conftest.dSYM: is a directory
rm: conftest.dSYM: is a directory
no
checking for dladdr... yes
checking for gettid... no
checking for lchown... yes
checking for setpriority... yes
checking for strerror... yes
checking for syscall... yes
checking for secure_getenv... no
checking for __secure_getenv... no
checking whether gcc -arch arm64 accepts -pthread... yes
configure: creating ./config.status
config.status: creating Makefile
config.status: creating config/Makefile
config.status: creating config/autoconf.mk
config.status: creating config/nsprincl.mk
config.status: creating config/nsprincl.sh
config.status: creating config/nspr-config
config.status: creating config/nspr.pc
config.status: creating lib/Makefile
config.status: creating lib/ds/Makefile
config.status: creating lib/libc/Makefile
config.status: creating lib/libc/include/Makefile
config.status: creating lib/libc/src/Makefile
config.status: creating lib/tests/Makefile
config.status: creating pkg/Makefile
config.status: creating pr/Makefile
config.status: creating pr/include/Makefile
config.status: creating pr/include/md/Makefile
config.status: creating pr/include/obsolete/Makefile
config.status: creating pr/include/private/Makefile
config.status: creating pr/src/Makefile
config.status: creating pr/src/io/Makefile
config.status: creating pr/src/linking/Makefile
config.status: creating pr/src/malloc/Makefile
config.status: creating pr/src/md/Makefile
config.status: creating pr/src/md/unix/Makefile
config.status: creating pr/src/memory/Makefile
config.status: creating pr/src/misc/Makefile
config.status: creating pr/src/threads/Makefile
config.status: creating pr/tests/Makefile
config.status: creating pr/tests/dll/Makefile
config.status: creating pr/src/pthreads/Makefile
config.status: executing default commands
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C config export
gcc -o now.o -c -g -fPIC -arch arm64 -mmacosx-version-min=10.11 -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.1.sdk -DXP_UNIX now.c
gcc  now.o   -o now
ld: warning: ignoring file now.o, building for macOS-x86_64 but attempting to link with file built for unknown-arm64
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

</details>

**Comment by memsharded** [comment]
Hi @timblechmann 

I am trying to understand what is happening here. 
I would say that your ``nspr`` package is used as a ``build_require`` in the build context? If it is a build require, it should be in the "build" context and it shouldn't get the ``arm`` architecture flag, but the x86_64 one, and it shouldn't fail. Is it possible that it is not a build-require?

**Comment by timblechmann** [comment]
@memsharded, the nspr build system builds a tool (`now`) to obtain the timestamp and use this to generate sources, which are the compiled into the library. so the tool itself needs to be compiled for the host architecture.
the problem: if `-arch arm64` is passed as cflag, this flag will also be used to to compile the `now` tool, which is supposed to run in the "build context".

`nspr` actually compiles fine in the x86/arm64 cross toolchain by configuring with `./configure --build=x86_64-apple-darwin --host=aarch64-apple-darwin`. afaict #8118 helps some packages (where autotools do not handle apple silicon out of the box), but harms others like `nspr` (where apple silicon is built into autotools)

**Comment by jgsogo** [comment]
When the same Conan package is packaging a tool and a library, like `protobuf` (it contains the `protoc` tool and the `protobuf` library) we need in our workspace two packages, we need Conan to retrieve the package for out _host_ context to link with the library, and the package for our _build_ context to run the tool in our build machine. These packages have different `package-id` so they need to appear twice in the graph.

To achieve this scenario, Conan accepts two profiles, `--profile:host` and `--profile:build`. When two profiles are provided Conan will split the graph into a _host_ context and a _build_ context, the first one with the `requirements` and the second one with the `build_requirements`. 

It looks like that in this scenario you need to add `nspr` as a `requirement` and also as a `build_requirement` to your consumer recipe. Conan will propagate _linking_ information from the _host_ context (includes, libraries) and the environment from the _build_ context, typically the PATH to executables. [+info here](https://docs.conan.io/en/latest/devtools/build_requires.html?highlight=build_requires#build-and-host-contexts)

Recipes that are involved in cross-compiling scenarios are starting to be adapted to this approach which is closer to the Conan v2 defaults.

Have you tried your project in a (simpler) cross-compiling scenario? are you using two profiles?

**Comment by timblechmann** [comment]
> When the same Conan package is packaging a tool and a library

this is not the case with `nspr`: the process is roughly the following:
* compile `now`
* run `now` to generate source
* compile source into the library
* consume the library

especially: the tool is *not* packaged and *not* called by downstream code, but it's called by the nspr built itself.

a vanilla nspr configure behaves well:
```
../configure '--with-mozilla' '--disable-cplus' '--disable-64bit' '--enable-strip' '--enable-debug' --build=x86_64-apple-darwin --host=aarch64-apple-darwin
---->0----
/Applications/Xcode12.app/Contents/Developer/usr/bin/make -C config export
gcc -o now.o -c -DXP_UNIX ../../config/now.c
gcc  now.o   -o now
---->0----
gcc -arch arm64 -o unix.o -c      -Wall -fno-common -pthread -g -fPIC  -UNDEBUG -DDEBUG_tim -DPACKAGE_NAME=\"\" -DPACKAGE_TARNAME=\"\" -DPACKAGE_VERSION=\"\" -DPACKAGE_STRING=\"\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DMOZILLA_CLIENT=1 -DDEBUG=1 -DXP_UNIX=1 -DDARWIN=1 -DHAVE_BSD_FLOCK=1 -DHAVE_SOCKLEN_T=1 -DHAVE_POINTER_LOCALTIME_R=1 -DHAVE_CRT_EXTERNS_H=1 -DHAVE_DLADDR=1 -DHAVE_LCHOWN=1 -DHAVE_SETPRIORITY=1 -DHAVE_STRERROR=1 -DHAVE_SYSCALL=1 -DFORCE_PR_LOG -D_PR_PTHREADS -UHAVE_CVAR_BUILT_ON_SEM -D_NSPR_BUILD_ -I../../../../dist/include/nspr -I../../../../../pr/include -I../../../../../pr/include/private  ../../../../../pr/src/md/unix/unix.c
gcc -arch arm64 -o unix_errors.o -c      -Wall -fno-common -pthread -g -fPIC  -UNDEBUG -DDEBUG_tim -DPACKAGE_NAME=\"\" -DPACKAGE_TARNAME=\"\" -DPACKAGE_VERSION=\"\" -DPACKAGE_STRING=\"\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DMOZILLA_CLIENT=1 -DDEBUG=1 -DXP_UNIX=1 -DDARWIN=1 -DHAVE_BSD_FLOCK=1 -DHAVE_SOCKLEN_T=1 -DHAVE_POINTER_LOCALTIME_R=1 -DHAVE_CRT_EXTERNS_H=1 -DHAVE_DLADDR=1 -DHAVE_LCHOWN=1 -DHAVE_SETPRIORITY=1 -DHAVE_STRERROR=1 -DHAVE_SYSCALL=1 -DFORCE_PR_LOG -D_PR_PTHREADS -UHAVE_CVAR_BUILT_ON_SEM -D_NSPR_BUILD_ -I../../../../dist/include/nspr -I../../../../../pr/include -I../../../../../pr/include/private  ../../../../../pr/src/md/unix/unix_errors.c
gcc -arch arm64 -o uxproces.o -c      -Wall -fno-common -pthread -g -fPIC  -UNDEBUG -DDEBUG_tim -DPACKAGE_NAME=\"\" -DPACKAGE_TARNAME=\"\" -DPACKAGE_VERSION=\"\" -DPACKAGE_STRING=\"\" -DPACKAGE_BUGREPORT=\"\" -DPACKAGE_URL=\"\" -DMOZILLA_CLIENT=1 -DDEBUG=1 -DXP_UNIX=1 -DDARWIN=1 -DHAVE_BSD_FLOCK=1 -DHAVE_SOCKLEN_T=1 -DHAVE_POINTER_LOCALTIME_R=1 -DHAVE_CRT_EXTERNS_H=1 -DHAVE_DLADDR=1 -DHAVE_LCHOWN=1 -DHAVE_SETPRIORITY=1 -DHAVE_STRERROR=1 -DHAVE_SYSCALL=1 -DFORCE_PR_LOG -D_PR_PTHREADS -UHAVE_CVAR_BUILT_ON_SEM -D_NSPR_BUILD_ -I../../../../dist/include/nspr -I../../../../../pr/include -I../../../../../pr/include/private  ../../../../../pr/src/md/unix/uxproces.c
```

but conan tries to outsmart the apple silicon support in `autoconf` by adding `-arch arm64` explicitly. there may be good reasons to do this for some packages, but it breaks others like `nspr`

**Comment by SSE4** [comment]
I have reproduced an issue
this break for me as well (now conan, vanilla nspr 4.29):
```
#!/usr/bin/env bash

pushd nspr-4.29/nspr

export CFLAGS="-arch arm64"
./configure '--with-mozilla' '--disable-cplus' '--disable-64bit' '--enable-strip' '--enable-debug' --build=x86_64-apple-darwin --host=aarch64-apple-darwin
make clean
make

popd
```

```
gcc -o now.o -c -arch arm64 -DXP_UNIX now.c
gcc  now.o   -o now
ld: warning: ignoring file now.o, building for macOS-x86_64 but attempting to link with file built for unknown-arm64
Undefined symbols for architecture x86_64:
  "_main", referenced from:
     implicit entry/start for main executable
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [now] Error 1
make: *** [export] Error 2
```
I have found similar issue https://bugs.chromium.org/p/chromium/issues/detail?id=200306
as far as I understand, `CFLAGS` shouldn't leak into now.c compilation.

**Comment by SSE4** [comment]
I am reading https://github.com/NixOS/nixpkgs/issues/50656 and few other issues on bugzilla:
- https://bugzilla.mozilla.org/show_bug.cgi?id=1230573
- https://bugzilla.mozilla.org/show_bug.cgi?id=742033
- https://bugzilla.mozilla.org/show_bug.cgi?id=376256
- https://bugzilla.mozilla.org/show_bug.cgi?id=817622
it seems like NSPR configure is messed up.
TLDR: they messed up with host/build/target, and in NSPR target=host, host=build 🤯 

the following works (more or less)
```
 > ./configure '--with-mozilla' '--disable-cplus' '--disable-64bit' '--enable-strip' '--enable-debug' '--prefix=/Users/sse4/.conan/data/nspr/4.29/_/_/package/dedc51991a586e8d7ffa22d51136e70e21a48c76' '--bindir=${prefix}/bin' '--sbindir=${prefix}/bin' '--libexecdir=${prefix}/bin' '--libdir=${prefix}/lib' '--includedir=${prefix}/include' '--oldincludedir=${prefix}/include' '--datarootdir=${prefix}/share' --host=x86_64-apple-darwin --target=aarch64-apple-darwin
```

right away, I am not sure if we need to fix it in conan client itself. I'd better apply a fix to recipe (NSPR).

**Comment by jgsogo** [comment]
Back to `conan-center-index`, if the build-system of the library is broken, it is something to patch in the recipe (**submit the patch upstream too**). Thanks!
